### ISSUE TITLE
Writing a setup.py is Hard

### ISSUE THREAD

**Issue by dstufft** [pypa/packaging-problems#1]
Today most users tend to just cargo cult a setup.py around, never understanding most of the directives that actually go into one.

Solutions:
- [ ] Document Creating a New Package From Scratch
- [ ] "Package Init", Create a New Package Template?

Some of this already is starting to exist with the packaging guide's page on creating a package ([See here](https://python-packaging-user-guide.readthedocs.org/en/latest/packaging.html)), but that is incomplete.

**Comment by hickford** [comment]
I'd love an `pip init` command analogous to `npm init` https://www.npmjs.org/doc/cli/npm-init.html

> This will ask you a bunch of questions, and then write a package.json for you.

**Comment by phildini** [comment]
I, too, would like something like `pip init` analogous to `npm init`. Is there anything more to push for, here? I'm a little sad that the last comment was 2015.

**Comment by merwok** [comment]
Well pip is the installer, not the build tool.  But I think that flit has an init command!

**Comment by njsmith** [comment]
@phildini The challenge there is that any package init tool is necessarily going to be pretty strongly opinionated, and people have a lot of conflicting opinions, plus the best practices are in a state of flux... but `pip init` would be a single thing that had to work for everyone, and probably can't change substantially once it's released. So you have to answer questions like: Should it generate a skeleton for an application or for a library? Pipfile or requirements.txt? Do you use tox? Do you provide some test harness, and if so, do you integrate it into setup.py? (Some people think this is a great idea, some people hate it.) Can you even assume that people want a setup.py at all, given that we're [working on moving away from them entirely](https://www.python.org/dev/peps/pep-0517/)? I can imagine that in a few years a lot of these questions will have settled down somewhat, but right now putting something into pip feels premature.

The most popular solution right now seems to be cookiecutter, and the nice thing about it is that it doesn't have to be one-size-fits-all.

So one thing that might make sense as a short-term improvement would be to advertise cookiecutter at https://packaging.python.org/tutorials/distributing-packages/

**Comment by ncoghlan** [comment]
While I like the *idea* of cookiecutter, I've found it personally intimidating whenever I've tried to use it, as the popular templates tend to push you strongly in the direction of doing things *right* (separate documentation that actively encourages contributions, testing across multiple Python versions with tox, etc). That's a really good thing for professional use, but for personal use it pushed me into "Ugh, this is too much work adding things that I don't personally need, so I'm just going to skip publishing anything to PyPI, and keep this project solely for my own use direct from source control".

`flit init` does have an interactive setup, and we're working towards getting that to a point where we can recommend it unreservedly to new publishers.  That falls into the category of tools that don't expose a human editable setup.py at all though, so the key sticking point to date has been flit's inability to publish sdists that are backwards compatible with widely deployed installation tools (which expect to see a setup.py file).

Still, a combined recommendation of `flit init` (for "just the basics, thanks") and `cookicutter` (for "I want to do things differently from the way flit does them") could work.

**Comment by phildini** [comment]
Fascinating! I had no idea `flit` existed. I'm pretty concerned about recommending _more_ tools to our already-complex toolchain.

Right now, we officially recommend _at least_ `pip` and `setuptools`, but we strongly encourage `twine`, and we're thinking about encouraging `flit`. Having to be aware of 3+ tools to create and upload a python package seems like a lot? (not even getting into the issue of _getting_ `pip`, etc).

I agree that cookiecutter can be intimidating, and I'm honestly not a fan of "download this thing from `pip` and then point it at an arbitrary github repo", but given that `flit` doesn't make a `setup.py`, how would people feel about a PR recommending `cookiecutter`? Based on feedback here and in freenode#pypa, it seems about 50/50.

**Comment by dstufft** [comment]
fwiw flit is effectively a replacement for setuptools in the toolchain, not an additional thing.

**Comment by njsmith** [comment]
@ncoghlan I suppose one could make a "just the basics" cookiecutter and publicize it here. Not sure who would be volunteering to do this though.

**Comment by pfmoore** [comment]
I agree with @ncoghlan, cookiecutter is quite intimidating in its basic form. However, it would be possible to recommend cookiecutter plus a specific project template that we recommend as a basic form to use. The difficulty is (as always) that there is no consensus on what form a recommended template should take.

(Hmm, I note that @njsmith just said much the same thing :-))

**Comment by ncoghlan** [comment]
Right, the idea with `flit` is for it to replace the `setuptools + twine` part of the current recommendations, so that the baseline recommendations become:

* use `pip` for package installation
* use `flit` for publishing pure Python packages

`setuptools` & `twine` won't be going anywhere (especially while https://github.com/takluyver/flit/issues/119 is still open), but their minimal level of complexity is quite high, whereas `flit` is more opinionated about how certain things should work (e.g. always getting your source file listing from your version control system, rather than having a separate manifest definition).

**Comment by merwok** [comment]
> The challenge there is that any package init tool is necessarily going to be pretty strongly opinionated, and people have a lot of conflicting opinions, plus the best practices are in a state of flux

I like that njs said «challenge», not «problem»! Compared to 2018, now that we have pyproject.toml unlocking build backends and frontends, the context is quite different:
- with the various build backends, people may not even have to write a setup.py! issue closed 😉
- setuptools still exists, with its docs emphasizing setup.cfg then pyproject.toml, and not emphasizing custom commands (or dist class 🙀) or generally custom code in setup.py
- unexpectedly, multiple build frontends exist, many of which you could call workflow tools, with their own init command and config system

So from points 1 and 2, this issue could be closed (in its narrow sense: it’s easier or obsolete now), but point 3 shows that we’re in an interesting time with innovation and competition between flit+pip, hatch, pixi, poetry… With useful convergence on standards, and fruitful divergence on extra features and user experience.

OK to close this @woodruffw ?

**Comment by woodruffw** [comment]
I think so!
